### DISCUSSION TITLE
Rate Limiting

### DISCUSSION THREAD

**Discussion by andriusign** [directus/directus#7134] [Archive]
I have started getting some 429 errors from API (running directus:9.0.0-rc.85 Dockerized).
![image](https://user-images.githubusercontent.com/14263959/127896051-b19e2bf2-1b51-4bdb-b9bd-cb27e7769572.png)

This happens randomly, some images get loaded, some not, users started reporting they cannot login. Read Directus documentation at https://docs.directus.io/guides/api-config/#rate-limiting

My rate limiter settings are:
RATE_LIMITER_ENABLED="true"
RATE_LIMITER_POINTS="25"
RATE_LIMITER_DURATION="1"
RATE_LIMITER_STORE="memory"

This equals to 25 requests per 1 second, which is more than recommended 10 req per 5 s in documentation, if I understand correctly. At the moment there are registered 4913 users (not sure if they all hit API at the same time though :smiley: )

1. Maybe someone already encountered similar problems and you found solutions?
2. What would be recommendations when limiting rate?
3. Where is better to store rate limiter information (redis vs memory)?
4. How is rate limiter tracking unique users (PP, headers, something else)? Is there a difference between different roles (public vs authorized)?
5. How much can Directus hold? If I set limiter to 200 requests per 1 sec would that be of any impact to Directus?

My server specs are:
KVM server VPS 1 x 2.60 GHz (processor load ~15-20%), 4 GB RAM (~3 GB in use), 40 GB HDD, 100 Mbps broadband?

Checked metrics of network packets

![image](https://user-images.githubusercontent.com/14263959/127899996-7960b010-df85-4e6d-9051-0f009a8f6179.png)

From what I see at the peak time we had about 800 packets coming in each second. Does that mean I have to set rate limiter to handle 1000 requests per seconds to get no errors at all?

Any thoughts, suggestions or questions would be helpful.

**Comment by rijkvanzanten** [comment]
> Maybe someone already encountered similar problems and you found solutions?

This depends on your project and what you do with it. Can't say much around solutions there, as it's custom code outside of Directus. One quick improvement you could do is adding a CDN in front (of the assets endpoint), so you don't hit the Directus API as much. 

> What would be recommendations when limiting rate?

It depends on your system specs and how you're using Directus. The defaults (25/s) are very strict to support lighter systems as well. If you run a load-balanced cluster of Directus instances, this number can be way higher.

> Where is better to store rate limiter information (redis vs memory)?

Redis, memory doesn't support horizontal scaling, as it doesn't share the state with the other instances

> How is rate limiter tracking unique users (PP, headers, something else)? Is there a difference between different roles (public vs authorized)?

It's based on incoming IP address:

https://github.com/directus/directus/blob/561b81a0e59432b06637914d08add1b3afeaacc6/api/src/middleware/rate-limiter.ts#L20

> How much can Directus hold? If I set limiter to 200 requests per 1 sec would that be of any impact to Directus?

This totally depends on your system resources, what sort of data you're fetching, and what third party services you use for database/file storage.

**Reply by andriusign** [reply]
Thanks a lot for sharing your knowledge.
